Add support for Bitbucket Code Insights

goober opened the pull request:
This change will add support for Bitbucket Code Insights in favor of regular comments when available. It will fall back on the comments strategy when the Code Insights is not available (it is supported in version 5.15 and later).

goober commented:
@4n4n4s maybe you have some good input on this one?

4n4n4s commented on src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/bitbucket/insights/client/BitbucketClient.java:
```diff
@@ -0,0 +1,149 @@
+        if (request.getAnnotations().isEmpty()) {
+            return;
+        }
+        Request req = new Request.Builder()
```
According to documentation:

> A report cannot have more than 1000 annotations by default, however this property is congurable at an instance level. If the request would result in more than the maximum number of annotations being stored then the entire request is rejected and no new annotations are stored.

We should make sure that for bigger projects all the annotations are still shown by splitting it up into max 1000 annotation batches.

goober replied:
I have added logic for splitting the annotations in batches of 1000.
In addition I sort the issues by severity so if there are more than the max number of allowed annotations the ones with highest severity will be shown.

4n4n4s commented on src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/bitbucket/insights/BitbucketServerCodeInsightsPullRequestDecorator.java:
```diff
@@ -0,0 +1,182 @@
+                "SonarQube",
+                analysisDetails.getAnalysisDate().toInstant(),
+                analysisDetails.getDashboardUrl(),
+                "https://www.sonarqube.org/favicon-152.png",
```
suggestion deliver this image also in the static resources so that instances of SQ that might not be able to access the internet still have an icon displayed see `/common/bug.svg`

goober replied:
Fixed

4n4n4s commented on src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/bitbucket/insights/BitbucketServerCodeInsightsPullRequestDecorator.java:
```diff
@@ -0,0 +1,182 @@
+
+    private String toBitbucketType(String sonarqubeType) {
+        switch (sonarqubeType) {
+            case "SECURITY_HOTSPOT":
```
we could use RuleType eg. RuleType.SECURITY_HOTSPOT.name()

goober replied:
Fixed

4n4n4s commented on src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/bitbucket/insights/BitbucketServerCodeInsightsPullRequestDecorator.java:
```diff
@@ -0,0 +1,182 @@
+            return "LOW";
+        }
+        switch (severity) {
+            case "BLOCKER":
```
we could use Severity eg. Severity.BLOCKER

goober replied:
Fixed

4n4n4s commented on src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/bitbucket/insights/client/model/CreateReportRequest.java:
```diff
@@ -0,0 +1,70 @@
+    public String getResult() {
+        return result;
+    }
+}
```
no newline at EOF

goober replied:
Fixed

goober commented on src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/bitbucket/insights/client/model/CreateReportRequest.java:
```diff
@@ -0,0 +1,70 @@
+            @JsonProperty("data") List<ReportData> data,
+            @JsonProperty("details") String details,
+            @JsonProperty("title") String title,
+            @JsonProperty("vendor") String vendor,
```
Thanks, this has been fixed

goober commented:
@4n4n4s 

> Maybe the "code insights" implementation should completely replace the comments feature and we already should set it to deprecated as 5.15 of bitbucket server was realeased in 2018.

I agree with you, should I remove the comments implementation in the same PR or should we do it in a separate one? We have just upgraded to Bitbucket Server 7.1.1 and unfortunately, the comments with included images looks really awful.
![image](https://user-images.githubusercontent.com/124244/79767363-c40f1d00-8318-11ea-9f66-f89e5d79420c.png)

 
> Similar to all the other files in the project please also add the copyright and other info on top of all your added files.

Done

> Added some remarks to files, didn't have the time yet to make a test drive on my local instance. For example would be good to know if #149 is also fixed with the annotations.

I have manually verified that this is not an issue with the Code Insights API

mc1arke commented:
> should I remove the comments implementation in the same PR or should we do it in a separate one?

I'd remove it in this PR. The delegate implementations currently implement the Decorator interface but don't really have a need to, so removing the old implementation should allow you to collapse code together and simplify the whole implementation.

goober commented:
@mc1arke Thanks for the feedback. I have updated the pull request and removed the comments implementation all together. 

@4n4n4s since you did the initial implementation it would be great if you could take a second look so I have not forgotten to remove something.

mc1arke commented on src/test/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/bitbucket/BitbucketPullRequestDecoratorTest.java:
```diff
@@ -0,0 +1,156 @@
+        assertThat(actual.getResult()).isEqualTo("PASS");
+        assertThat(actual.getReporter()).isEqualTo("SonarQube");
+        assertThat(actual.getCreatedDate()).isBetween(Instant.now().minus(1, ChronoUnit.MINUTES), Instant.now());
+        assertThat(actual.getDetails()).isEqualTo("Quality Gate passed\n");
```
Use `System.lineSeparator()` rather than `\n` to prevent this failing on Windows.

mc1arke commented on src/test/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/bitbucket/BitbucketPullRequestDecoratorTest.java:
```diff
@@ -0,0 +1,156 @@
+    ArgumentCaptor<CreateReportRequest> reportCaptor;
+
+    @Captor
+    ArgumentCaptor<CreateAnnotationsRequest> annotationsCaptor;
```
Nitpick: these should really be `private`, and below the `static` fields. You've also mixed your annotation driven mocking/capturing with static mocking of `analysisDetails` and `unifyConfiguration`. I'd personally remove the `@Captor` and make this`ArgumentCaptor.for(CreateAnnotationsRequest.class)`.

mc1arke commented on src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/bitbucket/client/BitbucketException.java:
```diff
@@ -0,0 +1,50 @@
+                .map(ErrorResponse::getErrors)
+                .map(e -> e.stream()
+                        .map(ErrorResponse.Error::getMessage)
+                        .collect(Collectors.joining("\n")))
```
Use `System.lineSeparator()` rather than `\n`

mc1arke commented on src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/bitbucket/BitbucketServerPullRequestDecorator.java:
```diff
@@ -0,0 +1,235 @@
+            case BUG:
+                return "BUG";
+            default:
+                return "";
```
I'd throw in `IllegalStateException` rather than return an empty string given this would be an unexpected state.

mc1arke commented on src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/AnalysisDetails.java:
```diff
@@ -267,13 +279,12 @@ public String getAnalysisProjectKey() {
 
-
-    private List<QualityGate.Condition> findFailedConditions() {
+    public List<QualityGate.Condition> getFailedConditions() {
```
Any reason for renaming this?

mc1arke commented on src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/AnalysisDetails.java:
```diff
@@ -101,10 +101,22 @@ public String getCommitSha() {
     }
 
+    public Map<RuleType, Long> getRulesCount() {
+        return countRuleByType();
```
We not just make `countRuleByType` public, rather than introduce a method that just delegated directly to it?

goober replied:
No, I have reverted the change

goober commented:
@mc1arke I have updated the PR and squashed the commits according to your comments.

mc1arke commented:
@goober I've merged this with a couple of small modifications for the issues highlighted by Sonarqube and a few formatting tweaks,

marvin-w commented:
@goober @mc1arke While trying to get this to work for the cloud version I encountered a really strange behavior.

Is there a chance that this PR broke the overrides for project specific parameters for bitbucket? Only Project and Repository can be overrriden anymore since those come from the UnifiyConfiguration object. All other properties (like URL and token) can now only be defined globally. Individual project settings are no longer taken into account. Did someone test this?

mc1arke commented:
@marvin-w I've not specifically tested this so can't comment. Since I've just merged the Sonarqube 8.1 changes into master, `UnifyConfigration` wont exist any more. I'll therefore created a `release/1_3` branch if you want to fix any issues, or add features to the plugin supporting the current LTS version of Sonarqube.

The latest plugin releases will currently only support over-riding configuration in the Gitlab decorator since most configuration is now driven through Sonarqube APIs
